light-prover/prover/batch_append_circuit.go

```diff
@@ -0,0 +1,360 @@
+package prover
+
+import (
+	"fmt"
+	"light/light-prover/logging"
+	merkle_tree "light/light-prover/merkle-tree"
+	"light/light-prover/prover/poseidon"
+	"math/big"
+
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/reilabs/gnark-lean-extractor/v2/abstractor"
+)
+
+type BatchAppendCircuit struct {
+	// Public inputs
+	OldSubTreeHashChain frontend.Variable `gnark:",public"`
+	NewSubTreeHashChain frontend.Variable `gnark:",public"`
+	NewRoot             frontend.Variable `gnark:",public"`
+	HashchainHash       frontend.Variable `gnark:",public"`
+	StartIndex          frontend.Variable `gnark:",public"`
+
+	// Private inputs
+	Leaves              []frontend.Variable `gnark:",private"`
+	Subtrees            []frontend.Variable `gnark:",private"`
+	HashChainStartIndex frontend.Variable   `gnark:",private"`
+
+	BatchSize  uint32
+	TreeHeight uint32
+}
+
+func (circuit *BatchAppendCircuit) Define(api frontend.API) error {
+	if err := circuit.validateInputs(); err != nil {
+		return err
+	}
+
+	api.AssertIsEqual(circuit.HashChainStartIndex, 0)
```

Comment:
TODOs for a different pr:
- unify all circuits inputs into a single hashchain -> only one public input
- enable partial updates of `HashchainHash`  with `HashChainStartIndex` ( consider `HashchainHash` contains a batch of 5000 leaves but we want to use 5 proves to execute this update onchain)

Reply:
```suggestion
TODO:
	- unify all circuits inputs into a single hashchain -> only one public input
	- enable partial updates of `HashchainHash`  with `HashChainStartIndex`
	( consider `HashchainHash` contains a batch of 5000 leaves but we want to use 5 proves to execute this update onchain)
	api.AssertIsEqual(circuit.HashChainStartIndex, 0)
```